### PR TITLE
feat(utils/background-task-queue): run jobs in a Worker

### PR DESCRIPTION
**Pull request by sidvishnoi**
Summary:
https://github.com/marcoscaceres/respec.org/blob/7411a544853a27fcd83e11ed27520b5deb4c150e/utils/background-task-queue.ts#L110-L129

Server logs appear like this (no errors/stdout in logs):
image

And result of running job in the worker like this (stdout/stderr and summary in a log file):
image

### CODE REVIEW

File: utils/background-task-queue.ts [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+  type: "success" | "failure";
+  result: unknown;
+}
+
+if (!isMainThread) {
+  let task: (input: CallRequest["input"]) => Promise<unknown>;
+  parentPort.addListener("message", async (req: Request) => {
+    const { id } = req;
+
+    if (req.type === "init") {
```
Comment (marcoscaceres) [inline utils/background-task-queue.ts:33]: Nit: Maybe `switch` on req.type?

File: utils/background-task-queue.ts [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    }
+  });
+}
+
+class Lock {
+  private isLocked = false;
+  private emitter = new EventEmitter();
+
+  async acquire() {
+    while (this.isLocked) {
```
Comment (marcoscaceres) [inline utils/background-task-queue.ts:57]: Can this loop run more than once? It looks like it "unlock"s only on release()... if so: 
  
```suggestion
    if (this.isLocked) {
```

File: utils/background-task-queue.ts [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    this.isLocked = true;
+  }
+
+  release() {
+    this.isLocked = false;
+    this.emitter.emit("unlock");
+  }
+}
+
+class Logger {
```
Comment (marcoscaceres) [inline utils/background-task-queue.ts:69]: Logger looks fairly generic. Would it be worth putting into own file?

File: utils/background-task-queue.ts [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+    await new Promise<void>((resolve, reject) => {
+      this.worker.once("message", (response: Response) => {
+        if (response.id !== id) {
+          reject(new Error(`Failed to register worker module: ${modulePath}`));
+        } else {
+          if (response.result === "success") {
+            resolve();
+          } else {
+            reject(response.result);
```
Comment (marcoscaceres) [inline utils/background-task-queue.ts:220]: is result always a real `Error` in this case?

**Review by marcoscaceres — APPROVED** [review]
Looks good... couple of questions/nits...

File: utils/background-task-queue.ts (hunk 3 above)
Comment (sidvishnoi) [reply utils/background-task-queue.ts:69]: Not generic enough.

File: utils/background-task-queue.ts (hunk 4 above)
Comment (sidvishnoi) [reply utils/background-task-queue.ts:220]: Can be any value the function throws.

File: utils/background-task-queue.ts [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    const modulePath = this.modulePath;
+    const msg: OperationRequest = { id, type: "init", modulePath };
+    this.worker.postMessage(msg);
+
+    await new Promise<void>((resolve, reject) => {
+      this.worker.once("message", (response: Response) => {
+        if (response.id !== id) {
+          reject(new Error(`Failed to register worker module: ${modulePath}`));
+        } else {
+          if (response.result === "success") {
```
Comment (sidvishnoi) [inline utils/background-task-queue.ts:217]: Bug: this should be result.type

File: utils/background-task-queue.ts (hunk 1 above)
Comment (sidvishnoi) [reply utils/background-task-queue.ts:33]: I'm gonna pass this one 🙈🤺

File: utils/background-task-queue.ts (hunk 2 above)
Comment (sidvishnoi) [reply utils/background-task-queue.ts:57]: Ah. Now I remember. If we use `if`, multiple requests can start at once when the lock is released (the promise resolves at once). The `while` loop disallows it (I don't fully understand how, but that's how it works).

Comment (marcoscaceres) [reply utils/background-task-queue.ts:57]: updated suggestion to include what you said.

File: utils/background-task-queue.ts (hunk 4 above)
Comment (marcoscaceres) [reply utils/background-task-queue.ts:220]: Ok, might be worth wrapping in an `Error`, just to get the stack trace.

**Pull request merged**